### PR TITLE
Intoduce 3DS support (bis repetita)

### PR DESCRIPTION
**Pull request by faide**
Hi,

I have taken over the work of Brendan on the 3DSecure option for Braintree. I fixed the things you asked in the original PR (https://github.com/lionelbarrow/braintree-go/pull/170).

I am sure I am missing some tests and maybe other things. Please give me a heads-up if you're interested in seeing this work being merged upstream.

Cheers,
Florent

### CODE REVIEW

**Comment by faide** [comment]
I see the tests failing for the TestTransactionSkipAdvancedFraudChecks but I don't yet see if it is caused by my changes...

**Comment by faide** [comment]
Now it seems tests have failed in one case because the BT Gateway was too busy, not because of our code ...

**Comment by faide** [comment]
Yes! All (previous) tests pass

File: transaction_integration_test.go [hunk 1]
```diff
@@ -757,6 +757,9 @@ func TestTransactionSkipAdvancedFraudChecks(t *testing.T) {
 		},
 		Options: &TransactionOptions{
 			SkipAdvancedFraudChecking: true,
+			ThreeDSecure: &TransactionOptionsThreeDSecure{
+				Required: false,
+			},
```
Comment (leighmcculloch) [inline transaction_integration_test.go]: Can you move this into a new test, that tests specifically 3DS fields independent of the other risk data fields?

File: transaction_integration_test.go [hunk 2]
```diff
@@ -765,7 +768,16 @@ func TestTransactionSkipAdvancedFraudChecks(t *testing.T) {
...
+		// thus I expect we will receive a risk data struct
+		// but this struct should tell us the risk as NOT been evaluated...
+		// quote: The possible values of the risk decision are Not Evaluated, Approve, Review, and Decline
+		if tx2.RiskData.Decision != RiskDataNotEvaluated {
+			t.Fatalf("expected tx2.RiskData.Decision to be %q, got %q", RiskDataNotEvaluated, tx2.RiskData.Decision)
+		}
+	} else {
+		t.Fatal("expected tx2.RiskData to be set with Decision field to 'Not Evaluated' but was nil")
 	}
 }
```
Comment (leighmcculloch) [inline transaction_integration_test.go]: Can we add tests that test a 3DS transaction similar to in #170 here `TestTransaction3DSCreateTransactionAndSettleSuccess`, but using the logic in the Java SDK mentioned here https://github.com/lionelbarrow/braintree-go/pull/170/files#r123081324?

**Comment by faide** [comment]
The fake nonces we have are for cards that are not valid for 3ds and get rejected.

I think I'll need a PaymentMethodNonceGateway (https://developers.braintreepayments.com/reference/request/payment-method-nonce/create/java) to be able to obtain the needed nonces for the tests, from the credit card numbers that can be found in the documentation: https://developers.braintreepayments.com/guides/3d-secure/testing-go-live/java

This seem to be implemented in java and python but not in Go.

I do not find the XML reference to be able to implement this in Go... Do you have any pointers where I could find the XML format documentation?

**Comment by faide** [comment]
I tried replicating the 3ds nonce getter from the other SDK but always get a 404... so I downloaded the python SDK and ran the same test against the sandbox. I get the same 404 result... This is as if this method does not exist anymore

**Comment by faide** [comment]
the URL I post against is:
```
https://api.sandbox.braintreegateway.com:443/merchants/4ngqq224rnk6gvxh/three_d_secure/create_nonce/bvps8ckgst59crzs
```
I am wondering if the merchant account id (bvps8ckgst59crzs) we are using for integration test is 3dsecure enabled... Is there a way to get this information ?

**Comment by faide** [comment]
I contacted support and apparently I can create nonce from the credit card with another URL:

https://api.sandbox.braintreegateway.com:443/merchants/<mymerchantid>/payment_methods/<thecardtoken>/nonces

which is what the other SDK do with a PaymentMethodNonceGateWay

So I implemented such a GateWay but at the moment this does not work (404 error on the endpoint)

I contacted support on this matter also, but pushed the code so you could take a peek and maybe help me find some errors I may have done

**Comment by faide** [comment]
I am progessing on this. I may have something soon :)

**Comment by faide** [comment]
I have two tests failing locally (and in travis):

  - TestSaveCreditCardWithVenmoSDKSession
  - TestSaveCreditCardWithVenmoSDKPaymentMethodCode

They seem to be completly unrelated to my work and I did not touch around their use case. Do you have any idea what is going on with these two guys ?

**Comment by faide** [comment]
Fun fact those two tests passed on travis with commit 2297c62, but If I checkout this revision now and run the integration test again, they fail.

**Comment by faide** [comment]
At the moment the support tells me there is no pre defined nonces that can be provided for 3DS tests.

They also tell me that any test involving 3DS MUST use the client side, see support answer:

> For reference, Braintree does not provide pre-verified testing nonces for 3D Secure for any of our server SDKs. Instead, 3D Secure testing requires working with both the client-side and server-side, as the authentication flow is presented on the client-side and required for 3D Secure verification. During testing, you will always need to walkthrough the 3D Secure authentication workflow, even in the sandbox environment.

This means we are stuck with no unit / integration tests for this PR. We could possibly imagine using gopherJS to emulate a client side asking threeDSecure.verifyCard() but this begins to be a rather large endeavour and we would have to cope with all the quirks of crawling through the returned dom and emulate a user entering his CVV. And all this would be subject the the CVV being a fixed and predictable value for the said card (which is not given in the test cases of the sentinel PDF)

**Comment by faide** [comment]
I am trying to push them a little bit to action through [Stackoverflow](https://stackoverflow.com/questions/43659532/unit-testing-3dsecure/49380436#49380436), even though I have no big illusions in the outcome.
But I am an optimistic :) so I must ask all people interested to have proper integration tests for 3D Secure support to go and upvote my comment in this thread to help us get a chance to move them.

**Comment by leighmcculloch** [comment]
@faide Thanks for persevering to figure out how this can be tested. I'll take a look over the next few days and see if I can find a way to test it. 😄

**Comment by leighmcculloch** [comment]
I have an idea for how to test this and plan to take a stab at it in the next few weeks.

**Comment by leighmcculloch** [comment]
This PR has been closed because its changes have been merged into `master` with #253. Thanks @faide for your hard work on picking this up and collecting so much info about how we could and couldn't test this. ❤️

**Pull request closed without merging**